File: be/app/models/subscription.py
import enum
from sqlalchemy import Column, Integer, DateTime, ForeignKey, Enum
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.core.database import Base


class SubscriptionPlan(str, enum.Enum):
    FREE = "free"
    PRO = "pro"
    MAX = "max"


class SubscriptionStatus(str, enum.Enum):
    ACTIVE = "active"
    EXPIRED = "expired"
    CANCELLED = "cancelled"
    TRIAL = "trial"


class Subscription(Base):
    __tablename__ = "subscriptions"

    id = Column(Integer, primary_key=True)
    user_id = Column(
        Integer,
        ForeignKey("users.id", ondelete="CASCADE"),
        unique=True,
        nullable=False,
        index=True,
    )

    plan = Column(Enum(SubscriptionPlan), default=SubscriptionPlan.PRO, nullable=False)
    status = Column(Enum(SubscriptionStatus), default=SubscriptionStatus.ACTIVE, nullable=False)

    started_at = Column(DateTime(timezone=True), server_default=func.now())
    expires_at = Column(DateTime(timezone=True), nullable=True)

    previous_plan = Column(Enum(SubscriptionPlan), nullable=True)
    plan_changed_at = Column(DateTime(timezone=True), nullable=True)

    monitoring_suspended_at = Column(DateTime(timezone=True), nullable=True)

    # Relationships
    user = relationship("User", back_populates="subscription")
    history = relationship(
        "SubscriptionHistory",
        back_populates="subscription",
        cascade="all, delete-orphan",
    )

File: fe/lib/features/subscription/domain/entities/subscription.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'subscription.freezed.dart';

@freezed
class Subscription with _$Subscription {
  const factory Subscription({
    required int id,
    required int userId,
    required String plan,
    required String status,
    required DateTime startedAt,
    DateTime? expiresAt,
    String? previousPlan,
    DateTime? planChangedAt,
    DateTime? monitoringSuspendedAt,
  }) = _Subscription;

  const Subscription._();

  bool get isSuspended => monitoringSuspendedAt != null;
  bool get isActive => status == 'active';
}
